### PR TITLE
SetCartRuleProductRestrictions CQRS handling

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add cart rule product restrictions handling into SetCartRuleRestrictionsCommand/handler. Add additional behat tests.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | yno
| How to test?      | New command covered with behats and unused in any exposed UI page (new cart rules page is not yet finished) so it doesn't need manual testing.
| Fixed ticket?     | part of https://github.com/PrestaShop/PrestaShop/issues/10547
| Related PRs       | If theme, autoupgrade or other module change is needed to make this change work, provide a link to related PRs here.
| Sponsor company   | Your company or customer's name goes here (if applicable).

### CODE REVIEW

File: src/Adapter/CartRule/CommandHandler/SetCartRuleProductRestrictionsHandler.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public function handle(SetCartRuleProductRestrictionsCommand $command): void
+    {
+        $cartRuleId = $command->getCartRuleId();
+        $cartRule = $this->cartRuleRepository->get($cartRuleId);
+
+        $this->cartRuleRepository->setProductRestrictions($cartRuleId, $command->getRestrictionRuleGroups());
+
+        $cartRule->product_restriction = !empty($command->getRestrictionRuleGroups());
```
Comment (zuk3975) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleProductRestrictionsHandler.php]: as noticed in another PR (https://github.com/PrestaShop/PrestaShop/pull/32483), $cartRule->reduction_product is affected as well (at least in legacy code), but the two features (product restriction in conditions tab & specific product selection in actions tab) seems mixed up together and needs to be sorted out (restrictions defines rules when discount applies, reduction_product defines that the applicable discount should only be applied to that specific product and not whole order, so basically when there is a reduction_product it shouldn't need to be added to restrictions, because its already handled separately in FO.)
Example of current behavior - when you select specific product in actions tab, then that product appears in conditions -> product restrictions :sweat:

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php [hunk 2]
```diff
@@ -37,34 +37,63 @@
...
+        protected readonly CartRuleRepository $cartRuleRepository
     ) {
     }
 
     public function handle(SetCartRuleRestrictionsCommand $command): void
     {
-        $cartRuleId = $command->cartRuleId;
-        $cartRule = $this->cartRuleRepository->get($cartRuleId);
-        $restrictedCartRuleIds = $command->restrictedCartRuleIds;
+        if ($command->isEmpty()) {
```
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: I'm not sure this should be the responsibility of the command to indicate it's empty 🤔 The command is a pure DTO and `isEmpty` is not an information of a field being edited it's a state of the command itself

It makes more sense for the handler to check this IMHO

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php [hunk 3]
```diff
@@ -37,34 +37,63 @@
...
         // update cart_rule_restriction property for all the cart rules that have been affected
         foreach ($this->cartRuleRepository->getRestrictedCartRuleIds($cartRuleId) as $restrictedCartRuleId) {
             $affectedCartRule = $this->cartRuleRepository->get(new CartRuleId($restrictedCartRuleId));
-            $this->updateRestrictionProperty($affectedCartRule, $hasRestrictions);
+            $this->updateRestrictionProperty($affectedCartRule, 'cart_rule_restriction', $hasRestrictions);
         }
     }
 
-    private function updateRestrictionProperty(CartRule $cartRule, bool $hasRestrictions): void
+    private function updateRestrictionProperty(CartRule $cartRule, string $propertyName, bool $hasRestrictions): void
```
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: You're using this as a generic updater but it always modifies `$cartRule->cart_rule_restriction` I think you can keep this one for cart rule restriction only especially since you use it to update all the related cart rules

Besides you already update `$cartRule->product_restriction` in `setProductRestrictions` so you don't need to call `updateRestrictionProperty` after even if it worked as expected

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php [hunk 4]
```diff
@@ -37,34 +37,63 @@
...
+
+        $restrictedCartRuleIds = $command->getRestrictedCartRuleIds();
+        if (null !== $restrictedCartRuleIds) {
+            $this->setCartRuleRestrictions($cartRule, $restrictedCartRuleIds);
+            $this->updateRestrictionProperty($cartRule, 'cart_rule_restriction', !empty($restrictedCartRuleIds));
+        }
+        $productRestrictionGroups = $command->getProductRestrictionRuleGroups();
+        if (null !== $productRestrictionGroups) {
+            $this->setProductRestrictions($cartRule, $productRestrictionGroups);
+            $this->updateRestrictionProperty($cartRule, 'product_restriction', !empty($productRestrictionGroups));
```
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]:
```suggestion
```
The field update is already handled in `setProductRestrictions`

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php [hunk 5]
```diff
@@ -37,34 +37,63 @@
...
+            // no restrictions were modified
+            return;
+        }
+
+        $cartRule = $this->cartRuleRepository->get($command->cartRuleId);
+
+        $restrictedCartRuleIds = $command->getRestrictedCartRuleIds();
+        if (null !== $restrictedCartRuleIds) {
+            $this->setCartRuleRestrictions($cartRule, $restrictedCartRuleIds);
+            $this->updateRestrictionProperty($cartRule, 'cart_rule_restriction', !empty($restrictedCartRuleIds));
```
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]:
```suggestion
```
I believe this is already handled inside `setCartRuleRestrictions`

File: src/Adapter/CartRule/Repository/CartRuleRepository.php [hunk 6]
```diff
@@ -87,6 +90,162 @@ public function get(CartRuleId $cartRuleId): CartRule
...
+                    ])
+                    ->setParameter('productRuleGroupId', $productRuleGroupId)
+                    ->setParameter('type', $restrictionRule->type)
+                    ->execute()
+                ;
+
+                $productRuleId = $this->connection->lastInsertId();
+                $productRuleValues = [];
+                $checkedIds = [];
+                foreach ($restrictionRule->ids as $id) {
```
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php]: What does ID stand for? IS it productId, categoryId? Or can it be used for different entities?

File: src/PrestaShopBundle/Resources/config/services/adapter/cart_rule.yml [hunk 7]
```diff
@@ -66,6 +66,7 @@ services:
     autowire: true
     public: false
 
+
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/cart_rule.yml]: Can be cleaned I think

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    And I add product "product2" with following information:
+      | name[en-US] | T-shirt nr1 |
+      | type        | standard    |
+    And I add product "product3" with following information:
+      | name[en-US] | Shirt - Dom & Jquery |
+      | type        | standard             |
+
+  # @todo: create cart rules once in first scenario, so they are not recreated before every scenario.
+  #  This step is temporary and should be replaced by step "Given there is a cart rule..." in background
+  #  when following PR is merged: https://github.com/PrestaShop/PrestaShop/pull/32483
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature]: Since the refacto PR has been merged can you fix this part in this PR?

File: src/Adapter/CartRule/Repository/CartRuleRepository.php (hunk 6 above)
Comment (zuk3975) [reply src/Adapter/CartRule/Repository/CartRuleRepository.php]: yep, it can be category,product,attribute,supplier or manufacturer. Depends on $restriction->type

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php (hunk 2 above)
Comment (zuk3975) [reply src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: felt like a good shortcut :smile: I can get rid of it, at some "theoretical level" you are right :smile:

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php (hunk 3 above)
Comment (zuk3975) [reply src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]: yeah I mixed up 2 ideas that I had and forgot to clean half of it :bow:

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+        $command->setProductRestrictionRuleGroups([]);
+
+        $this->getCommandBus()->handle($command);
+        $this->restrictionCommandsByReference[$cartRuleReference] = $command;
+        unset($this->productRestrictionGroupsByReference[$cartRuleReference]);
+    }
+
+    /**
+     * @When I save product restrictions for cart rule :cartRuleReference
```
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php]: this "product" step can be removed. But if its the only thing to do then it shouldn't block this PR (I'll fix in upcoming PR)

Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php]: removed :heavy_check_mark:

File: src/Adapter/CartRule/Repository/CartRuleRepository.php (hunk 6 above)
Comment (jolelievre) [reply src/Adapter/CartRule/Repository/CartRuleRepository.php]: Can you please replace this public field with a getter instead (and same in other commands, although it can be done in a dedicated PR for others) This one is a bit too generic, so maybe something like:

```suggestion
                foreach ($restrictionRule->getEntityIds() as $entityId) {
```

File: src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php [hunk 10]
```diff
@@ -37,34 +37,54 @@
...
     {
-        $cartRuleId = $command->cartRuleId;
-        $cartRule = $this->cartRuleRepository->get($cartRuleId);
-        $restrictedCartRuleIds = $command->restrictedCartRuleIds;
+        if (null === $command->getRestrictedCartRuleIds() && null === $command->getProductRestrictionRuleGroups()) {
+            // no restrictions were modified
+            return;
+        }
+
+        $cartRule = $this->cartRuleRepository->get($command->cartRuleId);
```
Comment (jolelievre) [inline src/Adapter/CartRule/CommandHandler/SetCartRuleRestrictionsHandler.php]:
```suggestion
        $cartRule = $this->cartRuleRepository->get($command->getCartRuleId());
```

File: src/Adapter/CartRule/Repository/CartRuleRepository.php [hunk 11]
```diff
@@ -87,6 +90,162 @@ public function get(CartRuleId $cartRuleId): CartRule
...
+                ->values([
+                    'id_cart_rule' => $cartRuleId->getValue(),
+                    'quantity' => $restrictionRuleGroup->requiredQuantityInCart,
+                ])
+                ->execute()
+            ;
+
+            $productRuleGroupId = $this->connection->lastInsertId();
+
+            foreach ($restrictionRuleGroup->restrictionRules as $restrictionRule) {
```
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php]:
```suggestion
            foreach ($restrictionRuleGroup->getRestrictionRules() as $restrictionRule) {
```

File: src/Adapter/CartRule/Repository/CartRuleRepository.php [hunk 12]
```diff
@@ -87,6 +90,162 @@ public function get(CartRuleId $cartRuleId): CartRule
...
+                $this->connection->createQueryBuilder()
+                    ->insert($this->dbPrefix . 'cart_rule_product_rule')
+                    ->values([
+                        'id_product_rule_group' => ':productRuleGroupId',
+                        'type' => ':type',
+                    ])
+                    ->setParameter('productRuleGroupId', $productRuleGroupId)
+                    ->setParameter('type', $restrictionRule->type)
```
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php]:
```suggestion
                    ->setParameter('type', $restrictionRule->getType())
```

File: src/Adapter/CartRule/Repository/CartRuleRepository.php (hunk 6 above)
Comment (jolelievre) [reply src/Adapter/CartRule/Repository/CartRuleRepository.php]: Or `getRestrictedEntityIds` or `getRestrictedIds` I'm not sure

File: src/Adapter/CartRule/Repository/CartRuleRepository.php [hunk 13]
```diff
@@ -87,6 +90,162 @@ public function get(CartRuleId $cartRuleId): CartRule
...
+                'crpr.id_product_rule_group = crprg.id_product_rule_group'
+            )
+            ->where('crprg.id_cart_rule = :cartRuleId')
+            ->setParameter('cartRuleId', $cartRuleId->getValue())
+            ->execute()
+            ->fetchAllAssociative()
+        ;
+
+        // retrieve all rules based on cart rule
+        $rules = $this->connection->createQueryBuilder()
```
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php:178]: Seems like this query is quite similar with the previous one, they are both based on the same condition `'crprg.id_cart_rule = :cartRuleId'` and get data from related tables that are part of the joins in both queries

Isn't it possible to perform only one query and get all the needed data in one go?

File: src/Adapter/CartRule/Repository/CartRuleRepository.php [hunk 14]
```diff
@@ -87,6 +90,162 @@ public function get(CartRuleId $cartRuleId): CartRule
...
+                'crpr.id_product_rule_group = crprg.id_product_rule_group'
+            )
+            ->where('crprg.id_cart_rule = :cartRuleId')
+            ->setParameter('cartRuleId', $cartRuleId->getValue())
+            ->execute()
+            ->fetchAllAssociative()
+        ;
+
+        // retrieve all rule groups based on cart rule
+        $groups = $this->connection->createQueryBuilder()
```
Comment (jolelievre) [inline src/Adapter/CartRule/Repository/CartRuleRepository.php:194]: SAme question as the comment before, maybe even these three queries can be merged no?

File: src/Core/Domain/CartRule/Command/SetCartRuleRestrictionsCommand.php [hunk 15]
```diff
@@ -30,47 +30,83 @@
...
 use PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\CartRuleId;
+use PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\Restriction\RestrictionRuleGroup;
 
+/**
+ * Sets cart rule restrictions.
+ * Leaving property as NULL will cause no changes, however having an empty array will clear existing restrictions.
+ */
 class SetCartRuleRestrictionsCommand
 {
     public readonly CartRuleId $cartRuleId;
```
Comment (jolelievre) [inline src/Core/Domain/CartRule/Command/SetCartRuleRestrictionsCommand.php]: Should be replaced by a protected readonly and a getter instead

File: src/Core/Domain/CartRule/QueryResult/CartRuleRestrictionsForEditing.php [hunk 16]
```diff
@@ -28,11 +28,19 @@
...
-    //@todo: add other restrictions like country, carrier, group, product, shop (shop association might be handled separatly)
+    /**
+     * @todo: fill other restrictions when related commands are implemented
+     *
+     * @param int[] $restrictedCartRuleIds
+     * @param RestrictionRuleGroup[] $productRestrictionRuleGroups
+     */
     public function __construct(
-        public readonly array $restrictedCartRuleIds
+        public readonly array $restrictedCartRuleIds,
```
Comment (jolelievre) [inline src/Core/Domain/CartRule/QueryResult/CartRuleRestrictionsForEditing.php]: SAme here, we should use getters instead

File: src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRule.php [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        self::TYPE_MANUFACTURERS => self::TYPE_MANUFACTURERS,
+        self::TYPE_SUPPLIER => self::TYPE_SUPPLIER,
+    ];
+
+    /**
+     * @param string $type
+     * @param int[] $ids Ids of product, attributes, categories, manufacturers or suppliers, depending on restriction rule type
+     */
+    public function __construct(
+        public readonly string $type,
```
Comment (jolelievre) [inline src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRule.php]: To replace with protected fields and getters

File: src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRuleGroup.php [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+use PrestaShop\PrestaShop\Core\Domain\CartRule\Exception\CartRuleConstraintException;
+
+class RestrictionRuleGroup
+{
+    /**
+     * @param int $requiredQuantityInCart
+     * @param RestrictionRule[] $restrictionRules
+     */
+    public function __construct(
+        public readonly int $requiredQuantityInCart,
```
Comment (jolelievre) [inline src/Core/Domain/CartRule/ValueObject/Restriction/RestrictionRuleGroup.php]: SAme here to replace with getters

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php [hunk 19]
```diff
@@ -157,8 +153,87 @@ public function assertCartRuleError(string $error): void
...
+     * @Then cart rule :cartRuleReference should have the following product restriction rule groups:
+     *
+     * @return void
+     */
+    public function assertProductRestrictionGroups(string $cartRuleReference, TableNode $tableNode): void
+    {
+        $actualRestrictionGroups = $this->getCartRuleForEditing($cartRuleReference)
+            ->getConditions()
+            ->getRestrictions()
+            ->productRestrictionRuleGroups
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php]:
```suggestion
            ->getProductRestrictionRuleGroups()
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php [hunk 20]
```diff
@@ -157,8 +153,87 @@ public function assertCartRuleError(string $error): void
...
+        ;
+        $expectedDataRows = $tableNode->getColumnsHash();
+        Assert::assertCount(count($expectedDataRows), $actualRestrictionGroups, 'Unexpected product restriction groups count');
+
+        foreach ($expectedDataRows as $key => $expectedDataRow) {
+            $actualGroup = $actualRestrictionGroups[$key];
+            Assert::assertEquals(
+                $expectedDataRow['quantity'],
+                $actualGroup->requiredQuantityInCart,
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php]:
```suggestion
                $actualGroup->getRequiredQuantityInCart(),
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php [hunk 21]
```diff
@@ -157,8 +153,87 @@ public function assertCartRuleError(string $error): void
...
+                'Unexpected required quantity in cart in restriction group'
+            );
+            Assert::assertCount(
+                (int) $expectedDataRow['rules count'],
+                $actualGroup->restrictionRules,
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleAssertionFeatureContext.php]:
```suggestion
                $actualGroup->getRestrictionRules(),
```

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+     * @return void
+     */
+    public function saveRestrictionRules(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
+
+        if (isset($this->productRestrictionGroupsByReference[$cartRuleReference])) {
+            // set the restrictions that (if) were added step by step before
+            $command->setProductRestrictionRuleGroups($this->productRestrictionGroupsByReference[$cartRuleReference]);
+            unset($this->productRestrictionGroupsByReference[$cartRuleReference]);
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php:176]: Shouldn't `$this->restrictionCommandsByReference` also be cleaned once it's saved? To make sure no values from previous scenarios can remain in case a new scenario starts on the same cart rule

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+      | apply_to_discounted_products | false                  |
+      | discount_application_type    | order_without_shipping |
+    And cart rule "rule_free_shipping_1" should have the following properties:
+      | restricted cart rules |  |
+    And cart rule "rule_50_percent" should have the following properties:
+      | restricted cart rules |  |
+    And cart rule "rule_70_percent" should have the following properties:
+      | restricted cart rules |  |
+    And cart rule rule_free_shipping_1 should have no product restriction rules
+    And cart rule rule_50_percent should have no product restriction rules
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature:90]:
```suggestion
    And cart rule rule_50_percent should have no product restriction rules
    And cart rule rule_70_percent should have no product restriction rules
```
Third one should be checked as well no?

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+      | valid_from                   | 2022-01-01 11:00:00    |
+      | valid_to                     | 3001-01-01 12:00:00    |
+      | total_quantity               | 10                     |
+      | quantity_per_user            | 12                     |
+      | free_shipping                | false                  |
+      | code                         | rule_70_percent        |
+      | discount_percentage          | 70                     |
+      | apply_to_discounted_products | false                  |
+      | discount_application_type    | order_without_shipping |
+    And cart rule "rule_free_shipping_1" should have the following properties:
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature:83]: Should you clean them in case some additional scenarios are added? Like you did in the product restrictions feature file

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/CartRule/set_mixed_cart_rule_restrictions.feature:83]:
```suggestion
    And I clear all product restrictions for cart rule rule_free_shipping_1
    And I save all the restrictions for cart rule rule_free_shipping_1
    And I clear all product restrictions for cart rule rule_50_percent
    And I save all the restrictions for cart rule rule_50_percent
    And I clear all product restrictions for cart rule rule_70_percent
    And I save all the restrictions for cart rule rule_70_percent
    And cart rule "rule_free_shipping_1" should have the following properties:
```

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+      | type          | references    |
+      | manufacturers | graphicCorner |
+    And cart rule "rule_free_shipping_1" should have the following product restriction rule groups:
+      | groupReference | quantity | rules count |
+      | free_nr_1      | 4        | 1           |
+    And the cart rule restriction group "free_nr_1" should have the following rules:
+      | type          | references                 |
+      | manufacturers | studioDesign,graphicCorner |
+
+  Scenario: Restrict cart rule products by defining manufacturer matching rules
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature]:
```suggestion
  Scenario: Restrict cart rule products by defining supplier matching rules
```

File: tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+      | type      | references                          |
+      | suppliers | fashionSupplier,accessoriesSupplier |
+    And the cart rule restriction group "free_nr_3" should have the following rules:
+      | type          | references    |
+      | manufacturers | graphicCorner |
+
+  Scenario: Provide restriction group with empty list of restrictions
+    When I add a restriction for cart rule rule_50_percent, which requires at least 1 product, but I provide empty list of rules
+    Then I should get cart rule error about "empty restriction rules"
+    And cart rule rule_50_percent should have no product restriction rules
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature:284]: SHould we add some scenarios that edit the existing restrictions after initial one have been added? I believe since we're dealing with new cart rules and new group in each scenario you do edit them once but never twice 

And especially we should test that we can remove the previous rules, even remove them all and check that no restrictions are set anymore

Comment (zuk3975) [reply tests/Integration/Behaviour/Features/Scenario/CartRule/set_product_restrictions.feature:284]: added couple steps here. Should cover it.
https://github.com/PrestaShop/PrestaShop/pull/32556/commits/781363bf56e65d1d1474debf05563de354f0dbe2

File: tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+    /**
+     * @When I restrict cart rules for :cartRuleReference providing non-existing cart rules
+     *
+     * @param string $cartRuleReference
+     *
+     * @return void
+     */
+    public function restrictCartRulesProvidingNonExistingIds(string $cartRuleReference): void
+    {
+        $command = $this->getRestrictionsCommand($cartRuleReference);
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/CartRule/SetCartRuleRestrictionsFeatureContext.php]: In this endpoint, you execute the command right away so you shouldn't rely on the tmp saved command here because it could already contain some values inconsistent ith what the endpoint aims at doing

I think it would be safer to create a new dedicated command, set the invalid value and execute it, additionally, the persisted command could even be cleared.

TLDR; `getRestrictionsCommand` should ONLY be used in the endpoints that need stateful steps because of their complexity, this one can remain independent and stateless

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

**Review by boherm — APPROVED** [review]
LGTM, thanks @zuk3975!
